### PR TITLE
add status subresource for deployment

### PR DESCRIPTION
**Pull request by mikedanese**

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
Title confused me since it touches deployment

**Comment by mikedanese** [comment]
@derekwaynecarr oops, fixed.

**Comment by derekwaynecarr** [comment]
Thanks - could have sworn I previously did DaemonSet :-)

File: pkg/client/unversioned/deployment_test.go [hunk 1]
```diff
@@ -124,6 +124,27 @@ func TestDeploymentUpdate(t *testing.T) {
...
+	c := &testClient{
+		Request: testRequest{
+			Method: "PUT",
+			Path:   testapi.Extensions.ResourcePath(getDeploymentsResoureName(), ns, "abc") + "/status",
+			Query:  buildQueryValues(nil),
+		},
+		Response: Response{StatusCode: 200, Body: deployment},
+	}
+	response, err := c.Setup(t).Deployments(ns).UpdateStatus(deployment)
+	c.Validate(t, response, err)
```
Comment (nikhiljindal) [inline pkg/client/unversioned/deployment_test.go]: Change something in the status and verify that it was changed?
We should also verify that spec cannot be updated using UpdateStatus.
We should also have a similar test for ensuring that status cannot be updated using Update().

Comment (mikedanese) [inline pkg/client/unversioned/deployment_test.go]: ok. I think those would be tests in registery/deployment/etcd/. These client tests don't actually pass through registry IIUC.

File: pkg/registry/deployment/strategy.go [hunk 2]
```diff
@@ -61,6 +61,9 @@ func (deploymentStrategy) AllowCreateOnUpdate() bool {
 
 // PrepareForUpdate clears fields that are not allowed to be set by end users on update.
 func (deploymentStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	newDeployment := obj.(*extensions.Deployment)
```
Comment (nikhiljindal) [inline pkg/registry/deployment/strategy.go]: I see that we do the same thing in PrepareForCreate for other resources.
Ex: pkg/registry/ingress/strategy.go

**Comment by mikedanese** [comment]
@nikhiljindal I've addressed comments PTAL

File: pkg/registry/deployment/strategy.go [hunk 3]
```diff
@@ -46,6 +46,9 @@ func (deploymentStrategy) NamespaceScoped() bool {
 
 // PrepareForCreate clears fields that are not allowed to be set by end users on creation.
 func (deploymentStrategy) PrepareForCreate(obj runtime.Object) {
+	deployment := obj.(*extensions.Deployment)
+	deployment.Status = extensions.DeploymentStatus{}
+	deployment.Generation = 1
```
Comment (nikhiljindal) [inline pkg/registry/deployment/strategy.go]: I dont think we use Generation for deployment anywhere?

**Comment by nikhiljindal** [comment]
Thanks!
LGTM, modulo one comment.

**Comment by mikedanese** [comment]
@nikhiljindal Fixed!

**Comment by mikedanese** [comment]
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
LGTM, thanks!

**Pull request merged**